src/kubernetes/resource_version.rs

```diff
@@ -0,0 +1,92 @@
+//! Shared state bits for watch implementations.
+
+use k8s_openapi::apimachinery::pkg::apis::meta::v1::{ObjectMeta, WatchEvent};
+use k8s_openapi::Metadata;
+use std::ops::Deref;
+
+/// Resource version state in the context of a chain of watch requests.
+#[derive(Debug, Clone)]
+pub struct State(Option<String>);
+
+impl State {
+    /// Create a new resource version [`State`].
+    pub fn new() -> Self {
+        Self(None)
+    }
+
+    /// Update the resource version from a candidate obtained earlier.
+    pub fn update(&mut self, candidate: Candidate) {
+        self.0 = Some(candidate.0);
+    }
+
+    /// Reset the resource version. Use in case of a desync.
+    pub fn reset(&mut self) {
+        self.0 = None;
+    }
+
+    /// Get the current resource version value.
+    pub fn get(&self) -> Option<&str> {
+        self.into()
+    }
+}
+
+impl Deref for State {
```

Comment:
Some particularly pedantic people in the Rust ecosystem are very much against this being used for not-pointers. You may wish to respect their advice.

I don't have a particularly strong opionion here, but this is definitely a "magic-feeling" type so let's make sure it's a worthwhile usage.

Reply:
The goal here is make the interface feel like an inheritance. I've seen this used a lot for guards (as in lock guards etc) and opaque type wrappers - like this type.
I'd be happy to discuss this with the people that have problems with this approach, or read up an article is there is any. Maybe there are solid arguments against this pattern.

Reply:
That may be the case, but that `Deref` currently isn't used anywhere, and the `From/Into` below it are only used in the `State::get` method which could be implemented as `self.0.as_ref().map(String::as_str)`. So I don't see the use for `Deref` nor for `From`, yes they are nice, and yes they may be used in the future, but they aren't now, and they are easily addable later when the need arise.

Reply:
The thing is when I was building these units, I designed a certain API that goes with them, that's supposed to be meaningful. This is why I'm a bit hesitant to remove any parts of it based on them not being used currently. I feel like the current API is complete, in a sense that we shouldn't have a need to alter it later. Switching from this state to a state where we're lacking in some things that are easy to add later doesn't seem like a good transition to me.

---

I'd still be happy to know what people say about `Deref` and why they're against it. I did some digging, found a [stackoverflow question](https://stackoverflow.com/questions/45086595/is-it-considered-a-bad-practice-to-implement-deref-for-newtypes) about it, and even though the selected answer does not represent my view on the matter, the discussion there is on point.
I've also found the https://rust-lang.github.io/api-guidelines/predictability.html#only-smart-pointers-implement-deref-and-derefmut-c-deref which suggests that `Deref` should only be used for smart pointers, however, as I mentioned above, the very `std` uses them for lock guards for mutexes. Well, maybe they are also considered a kind of a smart pointer, so, fine.
There's another example in the `std` that might be relevant here: [std::num::NonZeroI8](https://doc.rust-lang.org/std/num/struct.NonZeroI8.html). This type doesn't implement a `Deref`, and it is a newtype - so that's up with that? In my view, it's still an absolutely fine place to use `Deref` - for the method coercion that it provides - but there's just a small problem there - it's that the pointers are bigger than the value itself, and the `i8` is `Copy`. Those two factors alter that sample case for our matter, in particular with them *not* implementing `Deref` and making `get` return by value. The bigger one - [std::num::NonZeroI128](https://doc.rust-lang.org/std/num/struct.NonZeroI128.html) is not bigger than a pointer (typically), but is still a `Copy`, it's `get` returns by-value, and it doesn't implement `Deref` either. But our underlying type is not `Copy`.
Another example from the `std` is [std::mem::ManuallyDrop](https://doc.rust-lang.org/std/mem/struct.ManuallyDrop.html#method.new). It implements `Deref` while not being a pointer at all - just like in our case here. Another interesting property is that it's a `#[repr(transparent)]`. We might also consider making our newtype `#[repr(transparent)]`! But that's a separate matter. So, there is a type at `std` that does pretty much what we do, and that's not a pointer (or a lock guard).

So, not to the bottom of it: why use `Deref` at all? The purpose of `Deref`, from the technical point of view, is to enable method coercion. I.e. it's so that we can do `state.is_some()`. I personally feel like `Deref` is a bad name for it, and that it should really be named differently, something like `RefCoerce` for that matter. But, the purpose of implementing `Deref` is to enable certain usage patterns for the type. The alternative here would be to either not offer this flexibility at all, and make users do `state.get().is_some()` (which doesn't achieve the goal though, as I want the newtype to feel mostly like the wrapped type), or to explicitly implement delegates to the wrapped type for every method we want there. I'd argue that explicit delegates are worse, because they arise even more controversial questions, require more effort to maintain, and, ultimately, are a more manual approach to produce the same result. So, I'd say the use of `Deref` for such cases is quite sensible.

Now, the question is why would I want the type to feel like a wrapped type in the first place? After all, as I explained above, the main value of `Deref` is that it effectively allows us to use methods from the wrapped type as if we had a value of that type. This is one of the properties of this newtype, that can be described as "I don't want to apply any restrictions to the inspection of the wrapped type, as long as they're read-only". Note I didn't implement `DerefMut`. This is, btw, a good way to express the usage rules you intend for the API - by implementing `Deref` and/or `DerefMut`.
One of the problems there is potential method collisions. In this case, I ensured there are no problematic ones.

I'd be happy to get feedback on my rationale here. I think it justifies the design of the API I used in this case.

I understand I may have missed something. I'd be happy if someone could find a concrete example (maybe from past experience) of when such design becomes problematic, because that would be something new for me and it'd affect my future designs. I'm not a big fan of altering code just because of democratic consensus if I feel that I don't share the consensus, but I'm super happy to learn new things and change my mind when I see a proper rationale.

Reply:
>  I feel like the current API is complete, in a sense that we shouldn't have a need to alter it later.

Fair enough. 

And honestly, I haven't read anything bellow the line. I feel that this quite important to you, but you can relax, the code isn't you. The explanation above the line is quite sufficient, and even if it wasn't, we would just ask what isn't clear to us. I'm making quite the assumptions, and maybe I'm overstepping boundaries, but I wanted to tell you that nevertheless.

Reply:
> And honestly, I haven't read anything bellow the line. I feel that this quite important to you, but you can relax, the code isn't you. The explanation above the line is quite sufficient, and even if it wasn't, we would just ask what isn't clear to us. I'm making quite the assumptions, and I'm overstepping boundaries, but I wanted to tell you that nonetheless.

Don't worry, I'm not taking it personally! Thx for the feedback and review, I appreciate and welcome your input, and input of everyone. 

I want to note that I meant "complete" not in a way that it's subject to change, or final version, but in a sense that it's "closed around itself", i.e. in a logical sense. The way I used it isn't clear, so I apologize.

The rest of that comment are my thoughts on `Deref` usage, cause I'm genuinely curious in improving my understanding of that area. Totally ok to skip if you're not interested :smile:

Reply:
Yeah, nrc was always the person who stopped me from using `Deref` for things like this.

Basically: It's an overhead question. `Deref` is magic and it's going to have some surprising interactions for folks as they work on the codebase. Using it on things like smart pointers makes them behave the way folks expect them to. I'm not sure you can say the same about `State`.

Given that this isn't used, and a `Deref` implementation is super easy to add, I'm not sure what the discussion is about. It sounds like we have dead code to remove.

> I feel like the current API is complete, in a sense that we shouldn't have a need to alter it later.

Mike, please. We all know the API will change over time. If you earnestly think that we should have a discussion about software evolution.

Reply:
Yeah, `Deref` is unused and is controversial. I'll remove it, makes the most sense.

> If you earnestly think that we should have a discussion about software evolution.

Honestly, I wouldn't mind that. Yet another one, but there's never too many.

Reply:
Sure! :) Feel free to set up a time. I kind of had a similar opinion for awhile but I've learnt that even boring otherwise unchanging code will eventually change. :) Compiler surface changes, processor architectures change, dependencies change, even requirements.

In this case, of particular note, is that the Kubernetes API is not stable and tends to change all the time, so it's highly unlikely this code will be future proofed against all future kubernetes versions in the next several decades!